[UI] Add Virtualise List for Volume Table

ChengYanJin opened the pull request:
**Component**: ui, tables

**Context**: 
With the use case of thousands of volumes, the table would meet performance issues when rendering them at once.
So instead of doing this, we just render a small subset of rows at a give time.

**Summary**:
Introduce _react-window_ and _react-virtualized-auto-sizer_  to realize the virtualize row. 
Adapt the interaction test

UI change: 
Replace the "Unknown" text in the **table** with a hyphen icon to reduce the noise that less important information brings.

**Acceptance criteria**: 
![image](https://user-images.githubusercontent.com/18453133/99789242-88446c00-2b22-11eb-8ebd-3a9a0a5e7540.png)
If we open the developer tool => Elements, we should see just a few of the `<tr/>` at the beginning, when we scroll down on the list, we can see the update of the list.
![image](https://user-images.githubusercontent.com/18453133/99790012-c68e5b00-2b23-11eb-84ee-33e0d7b64106.png)

---

<!-- Declare one or more issues to close once this PR gets merged -->

Closes: #2931

<!-- If you want to refer to an issue while not closing it, use:

See: #ISSUE_NUMBER

-->

ChengYanJin committed:
ui/table: Use the react-window to virtualize the lines

Autosizer automatically adjusts the width and height of a single child.

Refs: #2931

ui/table: Add flex for table row to fully use the space

Refs: #2931

ChengYanJin committed:
ui/table: Replace the text unknown by hyphen icon

Too many "Unknown" in the table seems extremly noisy and since the
"Unknown" information is rather less important, we should replace it by
an icon with tooltip.

Refs: #2931

ChengYanJin committed:
ui/package: Update the UI packages

- Upgrade the react-table version to get support for react (v17.0.1)
- Introduce react-window and react-virtualized-auto-sizer for the
  virtualize list

Refs: #2931

bert-e commented:
# Hello chengyanjin,



My role is to assist you with the merge of this
pull request. Please type `@bert-e help` to get information
on this process, or consult the [user documentation](
https://eve.devsca.com/github/scality/metalk8s/bert-e/doc/user).





*Status report is not available.*

bert-e commented:
# Branches have diverged



This pull request's source branch `improvement/add-virtualise-list-for-tables` has diverged from
`development/2.6` by more than 50 commits.

To avoid any integration risks, please re-synchronize them using one of the
following solutions:

* Merge `origin/development/2.6` into `improvement/add-virtualise-list-for-tables`
* Rebase `improvement/add-virtualise-list-for-tables` onto `origin/development/2.6`

Note: If you choose to rebase, you may have to ask me to rebuild
integration branches using the `reset` command.

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `improvement/add-virtualise-list-for-tables` into
`development/2.6`
* [w/2.7/improvement/add-virtualise-list-for-tables](https://github.com/scality/metalk8s/compare/development/2.7...w/2.7/improvement/add-virtualise-list-for-tables) will be merged into `development/2.7` 



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`
* `development/2.4`
* `development/2.5`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @jbertran

* @chengyanjin

* @slaperche-scality

* @thomasdanan

* @lucieleonard

* @gdemonet

* @teddyandrieux

* @nootal

* @monpote

* @tcarmet

* @sayf-eddine-scality

* @cmonfort

* @nicolast

* @carlito-scality

* @alexandre-allard-scality

bert-e commented:
# History mismatch



Merge commit #d8a6d63191aa0fd356244edda4009768c8f5f684 on the integration branch
`w/2.7/improvement/add-virtualise-list-for-tables` is merging a branch which is neither the current
branch `improvement/add-virtualise-list-for-tables` nor the development branch
`development/2.7`.

It is likely due to a rebase of the branch `improvement/add-virtualise-list-for-tables` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**

ChengYanJin committed:
ui/volumes: Fix the bugs cause too many API calls

For non-specified instance URL, such as `/volumes`,
we should automatically choose the first one in the list.

However, since we are not aware currectVolumeName in VolumePage
component, we fall into  update the URL which triger the
VolumePage re-renders. Cause the infinite API calls.

So I move this logic to VolumePageContent which solve the issue.

Refs: #2931

ChengYanJin committed:
ui/cypress: Fix the failed test caused by the table virtualization

Refs: #2931

ChengYanJin committed:
ui/tables: Replace all the html tags in table with div

Since <AutoSizer> is a <div/> so it breaks the table layout,
we need to use <div/> for all the HTML tags in table
(thead, tbody, tr, td...) and retrieve the defaullt styles by className.

Refs: #2931

thomasdanan reviewed (commented):
works well, just you have to mouse over very precisely on the hyphen to see the tooltip, may be the zone could be enlarged a bit

ChengYanJin committed:
ui/volumes: Increase the height of Unknown icon

So that the users don't need to hover on the hypen precisely

Refs: #2931

ChengYanJin commented:
> works well, just you have to mouse over very precisely on the hyphen to see the tooltip, may be the zone could be enlarged a bit

Done! Thanks for the feedback!

ChengYanJin commented:
/approve

bert-e commented:
# History mismatch



Merge commit #d8a6d63191aa0fd356244edda4009768c8f5f684 on the integration branch
`w/2.7/improvement/add-virtualise-list-for-tables` is merging a branch which is neither the current
branch `improvement/add-virtualise-list-for-tables` nor the development branch
`development/2.7`.

It is likely due to a rebase of the branch `improvement/add-virtualise-list-for-tables` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**





*The following options are set:* **approve**

ChengYanJin commented:
/reset

bert-e commented:
# Reset complete



I have successfully deleted this pull request's integration branches.






*The following options are set:* **approve**

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `improvement/add-virtualise-list-for-tables` into
`development/2.6`
* [w/2.7/improvement/add-virtualise-list-for-tables](https://github.com/scality/metalk8s/compare/development/2.7...w/2.7/improvement/add-virtualise-list-for-tables) will be merged into `development/2.7` 



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`
* `development/2.4`
* `development/2.5`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```







*The following options are set:* **approve**

bert-e commented:
# Build failed



The build for [commit](https://github.com/scality/metalk8s/commit/43bd056b91cecf0c8ad3956b02d76b189b4f2f3c) did not succeed in branch w/2.7/improvement/add-virtualise-list-for-tables.




*The following options are set:* **approve**

ChengYanJin committed:
ui/volumeTable: Wrap the word for each of the cell

Remove the unused styled for HTML elements

bert-e commented:
# History mismatch



Merge commit #9d38ea3d228a64369f3fad7f50ab396e33e5c700 on the integration branch
`w/2.7/improvement/add-virtualise-list-for-tables` is merging a branch which is neither the current
branch `improvement/add-virtualise-list-for-tables` nor the development branch
`development/2.7`.

It is likely due to a rebase of the branch `improvement/add-virtualise-list-for-tables` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**





*The following options are set:* **approve**

nicolas2bert commented on ui/src/components/VolumeListTable.js:
```diff
@@ -289,17 +263,86 @@ function Table({
     ?.isSortedDesc;
   useTableSortURLSync(sorted, desc, data);
 
+  const RenderRow = React.useCallback(
```
To lighten your component, you could move `RenderRow` to its own component.

ChengYanJin replied:
Hello @nicolas2bert, 

Thanks for taking a look at this PR!
Just to be clear, you suggested me to move `RenderRow` to `VolumeListTable` component in my case?

nicolas2bert replied:
I suggest you move `RenderRow` component out of `VolumeListTable` for clarity purposes.

ChengYanJin replied:
OK. Actually `RenderRow` is a part of the `Table` component. And I render `Table` component in the `VolumeListTable`.

I guess the confusing part is I put the `Table` component inside `VolumeListTable` file, which should be separated. 

But since only `VolumeListTable` would use this Table for the moment, I did not generalize the `Table` yet.

nicolas2bert replied:
Yes, moving `Table` component out of `VolumeListTable` would be another refacto.
I was talking about moving `RenderRow` out from `Table` and so being its own component.

nicolas2bert replied:
The idea is that the component is already pretty heavy (and might need refacto later as you mention), let's try not to add too much code on top of it.

nicolas2bert replied:
TableRow could even live in its own file to reduce the size of VolumeListTable.js

ChengYanJin replied:
Indeed, thanks!

ChengYanJin committed:
ui/volumes: Extract the TableRow component

Refs: #2931

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @jbertran

* @chengyanjin

* @slaperche-scality

* @thomasdanan

* @lucieleonard

* @gdemonet

* @teddyandrieux

* @nootal

* @monpote

* @tcarmet

* @sayf-eddine-scality

* @cmonfort

* @nicolast

* @carlito-scality

* @alexandre-allard-scality










*The following options are set:* **approve**

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `improvement/add-virtualise-list-for-tables` into
`development/2.7`
* [w/2.8/improvement/add-virtualise-list-for-tables](https://github.com/scality/metalk8s/compare/development/2.8...w/2.8/improvement/add-virtualise-list-for-tables) will be merged into `development/2.8` 



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`
* `development/2.4`
* `development/2.5`
* `development/2.6`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```







*The following options are set:* **approve**

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @jbertran

* @chengyanjin

* @slaperche-scality

* @thomasdanan

* @lucieleonard

* @gdemonet

* @teddyandrieux

* @nootal

* @monpote

* @tcarmet

* @sayf-eddine-scality

* @cmonfort

* @nicolast

* @carlito-scality

* @alexandre-allard-scality










*The following options are set:* **approve**

ChengYanJin committed:
ui/table: Extract the TableRow outside of VolumeListTable

Refs: #2931

JBWatenbergScality commented on ui/src/components/TableRow.js:
```diff
@@ -0,0 +1,98 @@
+`;
+
+const TableRow = (props) => {
+  const { row, style, rowClicked, volumeName, theme } = props;
```
I would rename `rowClicked` props to `onClick` to match standard event names

JBWatenbergScality replied:
Instead of passing volumeName I would introduce an isSelected prop that reflect in a clearer way if the row is selected or not

JBWatenbergScality replied:
In order to avoid props drilling of the theme you can rely on styled component which automatically map the theme as part of the props or use `useTheme` hook that retrieve current theme from styled-component internal context

JBWatenbergScality commented on ui/src/components/VolumeListTable.js:
```diff
@@ -151,6 +100,13 @@ const TooltipContent = styled.div`
   min-width: 60px;
 `;
 
+const UnknownIcon = styled.i`
```
This component is duplicated in table row, please consider exporting it instead if it is necessary

JBWatenbergScality commented on ui/src/components/VolumeListTable.js:
```diff
@@ -289,17 +246,31 @@ function Table({
+    ({ index, style }) => {
+      const row = rows[index];
+      prepareRow(row);
+      // Extract the TableRow component outside of Table component
```
This comment can be deleted

JBWatenbergScality commented on ui/src/components/VolumeListTable.js:
```diff
@@ -357,68 +334,48 @@ function Table({
-                {intl.translate('no_volume_found')}
-              </td>
-            </HeadRow>
+                No Volume
```
A translation seems to be missing here, maybe `{intl.translate('no_volume_found')}`

ChengYanJin committed:
ui/table: Pass the isSelected to the TableRow instead of volumeName

Refs: #2931

ChengYanJin committed:
ui/table: Rename the rowClicked to onClick for table

To match standard event names

Refs: #2931

ChengYanJin committed:
ui/table: Retrieve the common TableRow styled from its component

Refs: #2931

ChengYanJin committed:
ui/table: Remove the theme props which pass to the components

Since style component can map the theme as part of the props. we don't
need to pass the theme props to the component.

Refs: #2931

bert-e commented:
# Build failed



The build for [commit](https://github.com/scality/metalk8s/commit/9e46321277e6a3de94794ed020e10aa6e997fc6a) did not succeed in branch improvement/add-virtualise-list-for-tables.




*The following options are set:* **approve**

bert-e commented:
# In the queue



The changeset has received all authorizations and has been added to the
relevant queue(s). The queue(s) will be merged in the target development
branch(es) as soon as builds have passed.

The changeset will be merged in:


* :heavy_check_mark: `development/2.7`

* :heavy_check_mark: `development/2.8`



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`
* `development/2.4`
* `development/2.5`
* `development/2.6`



There is no action required on your side. You will be notified here once
the changeset has been merged. In the unlikely event that the changeset
fails permanently on the queue, a member of the admin team will
contact you to help resolve the matter.

*IMPORTANT*

Please do not attempt to modify this pull request.

* Any commit you add on the source branch will trigger a new cycle after the
  current queue is merged.
* Any commit you add on one of the integration branches will be **lost**.

If you need this pull request to be removed from the queue, please contact a
member of the admin team now.




*The following options are set:* **approve**

bert-e commented:
# 


I have successfully merged the changeset of this pull request
into targetted development branches:


* :heavy_check_mark: `development/2.7`

* :heavy_check_mark: `development/2.8`



The following branches have **NOT** changed:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`
* `development/2.4`
* `development/2.5`
* `development/2.6`



Please check the status of the associated issue None.

Goodbye chengyanjin.
